### PR TITLE
[SYCL][NFC] Fix static code analysis concerns

### PR DESCRIPTION
**Pull request by smanna12**
Found via a static-analysis tool:
Pointer may be dereferenced after it was positively checked for NULL

Inside GenerateCode() function:

if (FD->isInlineBuiltinDeclaration() && Fn) {} ----> Pointer 'Fn' checked for NULL here

if (SyclOptReport.HasOptReportInfo(FD)) {
    llvm::OptimizationRemarkEmitter ORE(Fn); ---> Pointer 'Fn' is dereferenced by passing argument 1 to function "OptimizationRemarkEmitter" 

This patch checks Pointer 'Fn' inside routine to resolve the bug.

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

### CODE REVIEW

File: clang/lib/CodeGen/CodeGenFunction.cpp [hunk 1]
```diff
@@ -1540,6 +1540,7 @@ void CodeGenFunction::GenerateCode(GlobalDecl GD, llvm::Function *Fn,
   StartFunction(GD, ResTy, Fn, FnInfo, Args, Loc, BodyRange.getBegin());
 
   SyclOptReportHandler &SyclOptReport = CGM.getDiags().getSYCLOptReport();
+  assert(Fn && "Function cannot be null");
   if (SyclOptReport.HasOptReportInfo(FD)) {
```
Comment (premanandrao) [inline clang/lib/CodeGen/CodeGenFunction.cpp:1544]: It looks like Fn could be nullptr entering this routine.  There are at least two places above that Fn is checked before it is used, and those checks look necessary to me from the comments.  Further, the routine StartFunction that is called at line 1540 dereferences Fn inside it, so at a minimum moving this assert before that routine is called makes more sense.

I wonder if a better fix may be to change the line below to:
```suggestion
  if (Fn && SyclOptReport.HasOptReportInfo(FD)) {
```
I am not a 100% sure.

Comment (smanna12) [reply clang/lib/CodeGen/CodeGenFunction.cpp:1544]: Thanks @premanandrao for the review. This seems to me a better fix for checking pointer "Fn" inside the routine. 
@elizabethandrews implemented this OptReport, what is opinion for this change? Thank you.

Comment (elizabethandrews) [reply clang/lib/CodeGen/CodeGenFunction.cpp:1544]: I like @premanandrao's suggestion

Comment (smanna12) [reply clang/lib/CodeGen/CodeGenFunction.cpp]: Thank you for the reviews. Yes, this seems to me better fix to check the pointer "Fn" inside the routine. I have updated patch.

**Review by erichkeane — APPROVED** [review]
LGTM!  Thanks!

**Review by elizabethandrews — APPROVED** [review]

**Comment by smanna12** [comment]
Thank you everyone for the reviews.

**Comment by smanna12** [comment]
This PR is ready. @dm-vodopyanov could you merge it? Thanks.

**Pull request merged**